Procedural vtables and fully customizeable wide pointers (#2967)
rust-lang/rfcs: text/0000-procedural-vtables.md

to invent their own wide pointers (and thus custom dynamically sized types).

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

In order to mark a trait (`MyTrait`) as using a custom vtable layout, you implement the `CustomUnsized` trait for `dyn MyTrait`.

```rust
trait MyTrait: SomeSuperTrait {
    fn some_fn(&mut self, x: u32) -> i32;
}

impl CustomUnsized for dyn MyTrait {
    type WidePointer = MyWidePointer;
    // details later in this RFC
}

```

`MyWidePointer` is the backing type that is going to be used for wide pointers to the given trait, so

* `&dyn MyTrait`
* `Box<dyn MyTrait>`
* `Arc<dyn MyTrait>`

and any other container types that can use wide pointers. Normally when unsizing from a concrete
pointer like `&MyStruct` to `&dyn MyTrait` a wide pointer (that essentially is
`(&MyStruct, &'static Vtable)`) is produced. This is currently done via compiler magic, but
further down in this section you can see how it theoretically could be done in user code.
Likely it will stay compiler magic out of compile-time performance reasons.

All `impl`s of `MyTrait` will now use `MyWidePointer` for generating the wide pointer.

> The `TraitDescription` struct describes the metadata like the list of methods and a tree

Author: Sorry, that was a leftover from a previous RFC version

of super traits.

You are actually generating the wide pointer, not a description of it.
Since your `impl`'s `from` function is being interpreted in the target's environment, all target specific information will match up.
Now, you need some information about the `impl` in order to generate your metadata (and your vtable).
You get this information directly from the type (the `T` parameter) by adding trait bounds on it.

As an example, consider the function which is what normally generates your metadata.

```rust
#[repr(C)]
struct Pointer<T> {
    ptr: *mut T,
    vtable: &'static VTable<T>,
}

/// If the `owned` flag is `true`, this is an owned conversion like
/// in `Box<T> as Box<dyn Trait>`. This distinction is important, as
/// unsizing that creates a vtable in the same allocation as the object
/// (like C++ does), cannot work on non-owned conversions. You can't just
/// move away the owned object. The flag allows you to forbid such
/// unsizings by triggering a compile-time `panic` with an explanation
/// for the user.
unsafe impl<T: MyTrait> const CustomUnsize<Pointer> for T {

>     fn unsize<const owned: bool>(ptr: *mut T) -> Pointer<T> {

Author: Idk, I haven't been able to come up with anything better so far

>     fn unsize<const owned: bool>(ptr: *mut T) -> Pointer<T> {

Reviewer: I'm worried this doesn't actually work. `ptr` may be owned but not point to the start of the allocation. For example:
```rust
use core::ops::CoerceUnsized;

struct MyWrapper<T: ?Sized> {
    some_data: u32,
    object: T,
}

impl<T: CoerceUnsized<U>, U> CoerceUnsized<MyWrapper<U>> for MyWrapper<T> {}

fn bad() -> Box<MyWrapper<dyn MyTrait>> {
    Box::new(MyWrapper {
        some_data: 42,
        object: MyTraitImplementor,
    })
    // Undefined behavior: dealloc is called with a pointer 4 bytes offset from the allocation
} 
```

I think this might be solved this by making the `CustomUnsize` impl optional. I originally thought it had to be required but I suppose it should be possible for a user to customize the `dyn Trait` type via `CustomUnsized` *without* allowing safe `Box::new(TraitImplementor) as Box<dyn Trait>` conversion via `CustomUnsize`.  So to do the C++ shenanigans you would instead specialize the unsizing to only sound cases like so:
```rust
pub fn boxed_class<T: MyTrait>(data: T) -> Box<dyn MyTrait> {
    let ptr = Box::into_raw(Box::new((mytrait_vtable::<T>(), data)));
    unsafe { Box::from_raw(std::mem::transmute(ptr)) }
}
```

>     fn unsize<const owned: bool>(ptr: *mut T) -> Pointer<T> {

Reviewer 2: For C++, I think you would need `Pin<Box<dyn MyTrait>>` since C++ assumes objects don't move.

>     fn unsize<const owned: bool>(ptr: *mut T) -> Pointer<T> {

Author: >  `ptr` may be owned but not point to the start of the allocation

The `CustomUnsize` impl gets a pointer to the thing being unsized, which in this case is `MyWrapper<i32>`, so this would all work out. The only time you don't have a pointer to the base is when you manually play with allocations, which may indeed be a problem (so if you have a custom `Box` that prepends with some padding).

>     fn unsize<const owned: bool>(ptr: *mut T) -> Pointer<T> {

Reviewer: > The CustomUnsize impl gets a pointer to the thing being unsized

Ah, that explains my confusion around the slice example. I'll have to think on the implications of it a bit.

        // value, but this uses promotion
        // (https://doc.rust-lang.org/stable/reference/destructors.html?highlight=promotion#constant-promotion),
        // so the value lives long enough.
        Pointer {
            ptr,
            vtable: &default_vtable::<T>(),
        }
    }
}

/// DISCLAIMER: this uses a `Vtable` struct which is just a part of the
/// default trait objects. Your own trait objects can use any metadata and
/// thus "vtable" layout that they want.
unsafe impl Unsized for dyn MyTrait {
    type WidePointer = Pointer;
    fn size_of(ptr: Pointer) -> usize {
        ptr.vtable.size
    }
    fn align_of(ptr: Pointer) -> usize {
        ptr.vtable.align
    }
}

impl Drop for dyn MyTrait {
    fn drop(&mut self) {
        // using a dummy concrete type for `Pointer`
        let ptr = transmute::<&mut dyn Trait, Pointer<()>>(self);
        unsafe {
            let drop = ptr.vtable.drop;
            drop(&raw mut ptr.ptr)
        }
    }
}

const fn default_vtable<T: MyTrait>() -> VTable<T> {
    // `VTable` is a `#[repr(C)]` type with fields at the appropriate
    // places.
    VTable {
        size: std::mem::size_of::<T>(),

>         align: std::mem::size_of::<T>(),

Reviewer 3: Should this be `std::mem::align_of::<T>()`?

`CStr` instead of having a `Deref` impl that converts.

```rust
pub trait CStr {}

impl CustomUnsized<CStrPtr> for dyn CStr {
    type WidePointer = *mut u8;
    fn size_of(ptr: *mut u8) -> usize {
        unsafe { strlen(ptr) }
    }
    fn align_of(_: *mut u8) -> usize {
        1
    }
}
```

## `[T]` as sugar for a `Slice` trait

We could remove `[T]` (and even `str`) from the language and just make it desugar to
a `std::slice::Slice` (or `StrSlice`) trait.

```rust
pub trait Slice<T> {}

#[repr(C)]
struct SlicePtr<T> {
    ptr: *mut T,
    len: usize,
}

// This impl must be in the `vec` module, to give it access to the `vec`
// internals instead of going through `&mut Vec<T>` or `&Vec<T>`.
impl<T> CustomUnsize<SlicePtr<T>> for Vec<T> {
    fn unsize<const owned: bool>(ptr: *mut Vec<T>) -> SlicePtr<T> {
        SlicePtr {
            ptr: vec.data,
            len: vec.len,
        }
    }
}

impl<T> CustomUnsized for dyn Slice<T> {
    type WidePointer = SlicePtr<T>;
    fn size_of(ptr: SlicePtr<T>) -> usize {
        ptr.len
    }
    fn align_of(_: SlicePtr<T>) -> usize {
        std::mem::align_of::<T>()
    }
}

impl Drop for dyn Slice<T> {
    fn drop(&mut self) {

>         let wide_ptr = transmute::<&mut dyn Slice<T>, SlicePtr<T>>(self);

Author: How would `MaybeUninit` help here? We need `drop_in_place` below anyway

        let mut data_ptr = wide_ptr.ptr;
        for i in 0..wide_ptr.len {
            std::ptr::drop_in_place(data_ptr);
            data_ptr = data_ptr.offset(1);
        }
    }
}
```

## C++ like vtables

> Most of the boilerplate is the same as with regular vtables.

Reviewer 4: Might be worth clarifying how this could coexist with the existing C++ vtables RFC.

        unsafe {
            // This can get a better impl, but for simplicity we reuse
            // the existing function.
            transmute(Array2::index(idx))
        }
    }
}
```

## Zero sized references to MMIO

Instead of having one type per MMIO register bank, we could have one
trait per bank and use a zero sized wide pointer format.

```rust
trait MyRegisterBank {
    fn flip_important_bit(&mut self);
}

unsafe impl CustomUnsized for dyn MyRegisterBank {
    type WidePointer = ();
    fn size_of(():()) -> usize {
        4
    }
    fn align_of(():()) -> usize {
        4
    }
}

impl MyRegisterBank for dyn MyRegisterBank {
    fn flip_important_bit(&mut self) {
        std::mem::volatile_write::<bool>(42, !std::mem::volatile_read::<bool>(42))
    }
}
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

When unsizing, the `<dyn MyTrait as CustomUnsize>::unsize` is invoked.
The only reason that trait must be
`impl const CustomUnsize` is to restrict what kind of things you can do in there, it's not
strictly necessary. This restriction may be lifted in the future.

> For all other operations, the methods on `<dyn MyTrait as CustomUnsized>` is invoked.

Reviewer 5: ```suggestion
For all other operations, the methods on `<dyn MyTrait as CustomUnsized>` are invoked.
```

    T,
    const IMPL: &'static std::vtable::TraitDescription,
    const _owned: bool,
>(*const T) -> (*const T, &'static VTable<{num_methods::<IMPL>()}>) {
    // We generate the metadata and put a pointer to the metadata into 
    // the field. This looks like it's passing a reference to a temporary
    // value, but this uses promotion
    // (https://doc.rust-lang.org/stable/reference/destructors.html?highlight=promotion#constant-promotion),
    // so the value lives long enough.
    (
        ptr,
        &default_vtable::<T, IMPL>() as *const _ as *const (),
    )
}

/// DISCLAIMER: this uses a `Vtable` struct which is just a part of the
/// default trait objects. Your own trait objects can use any metadata and
/// thus "vtable" layout that they want.
pub const fn custom_vtable<
    const IMPL: &'static std::vtable::TraitDescription,
>() -> std::vtable::DstInfo {
    let mut info = DstInfo::new();
    unsafe {
        // We supply a function for invoking trait methods.
        // This is always inlined and will thus get optimized to a single
        // deref and offset (strong handwaving happening here).
        info.method_id_to_fn_ptr(|mut idx, parents, meta| unsafe {

>             let meta = *(meta as *const (*const(), &'static Vtable<IMPL>));

Reviewer 6: should these be 

```suggestion
            let meta = *(meta as *const (*const(), &'static VTable<{num_methods::<IMPL>}>));
```

?

Author: yes

/// unsizings by triggering a compile-time `panic` with an explanation
/// for the user.
unsafe impl<const IMPL: &'static std::vtable::TraitDescription> const CustomUnsize for Pointer<IMPL> {
    fn from<T, const owned: bool>(ptr: *const T) -> Self {
        // We generate the metadata and put a pointer to the metadata into 
        // the field. This looks like it's passing a reference to a temporary
        // value, but this uses promotion
        // (https://doc.rust-lang.org/stable/reference/destructors.html?highlight=promotion#constant-promotion),
        // so the value lives long enough.
        Pointer {
            ptr,
            vtable: &default_vtable::<T, IMPL>(),
        }
    }
}

/// DISCLAIMER: this uses a `Vtable` struct which is just a part of the
/// default trait objects. Your own trait objects can use any metadata and
/// thus "vtable" layout that they want.
unsafe impl<const IMPL: &'static std::vtable::TraitDescription> CustomUnsized for Pointer<IMPL> {
    fn method_id_to_fn_ptr(
        self,
        method_index: usize,
        super_tree_path: usize,
    ) -> *const () {
        let mut table = IMPL;

>         for parent in parents {

Reviewer 7: where does `parents` come from?

Author: oops, copy pasta, thanks!

            // we don't support multi-parents yet
            assert_eq!(parent, 0);

>             idx += table.methods.len();

Reviewer 7: where is idx defined?

#[repr(C)]
struct CStrPtr {
    ptr: *const u8,
}

impl<
    T,
    const IMPL: &'static std::vtable::TraitDescription,
> CustomUnsized for CStrPtr {
    fn method_id_to_fn_ptr(
        self,
        method_index: usize,
        super_tree_path: usize,
    ) -> *const () {
        panic!("CStr has no trait methods, it's all inherent methods acting on the pointer")
    }
    fn size_of(self) -> usize {
        unsafe { strlen(self.ptr) }
    }
    fn align_of(self) -> usize {
        1
    }
    fn drop(self) {
        // Nothing to drop (just `u8`s) and we are not in charge of dealloc
    }
    fn self_ptr(self) -> *const () {
        self.ptr
    }
}
```

## `[T]` as sugar for a `Slice` trait

We could remove `[T]` (and even `str`) from the language and just make it desugar to
a `std::slice::Slice` (or `StrSlice`) trait.

```rust
#[custom_wide_pointer = "SlicePtr"]
pub trait Slice<T> {}

#[repr(C)]
struct SlicePtr<T> {
    ptr: *const T,
    len: usize,
}

impl<
    T,
    const IMPL: &'static std::vtable::TraitDescription,
> CustomUnsized for Slice<T> {
    fn method_id_to_fn_ptr(
        self,
        method_index: usize,
        super_tree_path: usize,
    ) -> *const () {

>         panic!("CStr has no trait methods, it's all inherent methods acting on the pointer")

Reviewer 8: Is `CStr` a typo here?

Author: yes

As an example, consider the function which is what normally generates your metadata.

```rust
#[repr(C)]
struct Pointer {
    ptr: *mut u8,
    vtable: &'static VTable,
}

/// For going from a `&SomeStruct<dyn Trait>` to a field of `SomeStruct`.
unsafe impl CustomProjection for Pointer {
    unsafe fn project(ptr: Pointer, offset: usize) -> *mut u8 {
        ptr.ptr.offset(offset),
    }
}

unsafe impl CustomProjectionToUnsized for Pointer {
    /// For going from `&SomeStruct<dyn Trait>` to the unsized field
    unsafe fn project_unsized(ptr: Pointer, offset: usize) -> Pointer {
        Pointer {
            ptr: ptr.ptr.offset(offset),
            vtable: ptr.vtable,
        }
    }
}

/// If the `owned` flag is `true`, this is an owned conversion like
/// in `Box<T> as Box<dyn Trait>`. This distinction is important, as
/// unsizing that creates a vtable in the same allocation as the object
/// (like C++ does), cannot work on non-owned conversions. You can't just
/// move away the owned object. The flag allows you to forbid such
/// unsizings by triggering a compile-time `panic` with an explanation
/// for the user.

> unsafe impl<T: MyTrait> const CustomUnsize<Pointer> for T {

Reviewer: This might just be the `Unsize` trait.

>     fn unsize<const owned: bool>(ptr: *mut T) -> Pointer {

Reviewer: Hmmm. This is really interesting. I suppose you could use some totally opaque type as a placeholder for the data pointer (maybe an `extern type`?) which would be swapped out at coercion time. That would also solve the issues with projection because Rust could to do whatever it wants with that particular field...

I'll have to think about it some more but I like that direction.

>     fn unsize<const owned: bool>(ptr: *mut T) -> Pointer {

Author: We don't actually need a type, we can create a dangling symbolic pointer (this is a value in mir interpreter!) and just use `*const T` for whatever `T` we desire. Any interaction with the pointer beyond copying it around will make CTFE complain to you at compile-time. Since this is a special case, we can actually not just tell you about "dangling pointer" but special case the diagnostic to be really helpful

ideas that could support `CString` -> `CStr` unsizing by allowing `CString` to implement
`CStr` instead of having a `Deref` impl that converts.

```rust
pub trait CStr {}

impl CustomUnsized<CStrPtr> for dyn CStr {
    type WidePointer = *mut u8;
    fn size_of(ptr: *mut u8) -> usize {
        unsafe { strlen(ptr) }
    }
    fn align_of(_: *mut u8) -> usize {
        1
    }
}
```

## `[T]` as sugar for a `Slice` trait

We could remove `[T]` (and even `str`) from the language and just make it desugar to
a `std::slice::Slice` (or `StrSlice`) trait.

```rust
pub trait Slice<T> {}

#[repr(C)]
struct SlicePtr {
    ptr: *mut u8,
    len: usize,
}

unsafe impl CustomProjection for SlicePtr {
    unsafe fn project(ptr: Pointer, offset: usize) -> *mut u8 {
        ptr.ptr.offset(offset)
    }
}

unsafe impl CustomProjectionToUnsized for SlicePtr {
    unsafe fn project_unsized(ptr: Pointer, offset: usize) -> Pointer {
        Pointer {
            ptr.ptr.offset(offset),
            ptr.len,
        }
    }
}

// This impl must be in the `vec` module, to give it access to the `vec`
// internals instead of going through `&mut Vec<T>` or `&Vec<T>`.

> impl<T> CustomUnsize<SlicePtr> for Vec<T> {

Author: I don't see what the advantage is beyond having a simpler implementation. We can't reduce the complexity of the trait system described in this RFC if we limit ourselves to simple to convert types.

> impl<T> CustomUnsize<SlicePtr> for Vec<T> {

Reviewer: It is just the semantics of Vec. Vec is not a slice, it manages a slice. This is clear if you think about the implications of a `&Struct<Vec<T>>` to `&Struct<[T]>` coercion.
 I suppose it isn't a big deal. I'd just like to see a use case where there is no simple work around to this RFC via `Deref`.

    fn unsize<const owned: bool>(ptr: *mut Vec<T>) -> SlicePtr {
        SlicePtr {
            ptr: vec.data as *mut _,
            len: vec.len,
        }
    }
}

impl<T> CustomUnsized for dyn Slice<T> {
    type WidePointer = SlicePtr;
    fn size_of(ptr: SlicePtr) -> usize {
        ptr.len * std::mem::size_of::<T>()
    }
    fn align_of(_: SlicePtr) -> usize {
        std::mem::align_of::<T>()
    }
}

impl Drop for dyn Slice<T> {
    fn drop(&mut self) {
        unsafe {
            let wide_ptr = transmute::<&mut dyn Slice<T>, SlicePtr<T>>(self);

>             let mut data_ptr = wide_ptr.ptr;
>             for i in 0..wide_ptr.len {
>                 std::ptr::drop_in_place(data_ptr);
>                 data_ptr = data_ptr.offset(1);
>             }

Reviewer: ```suggestion
            let mut data_ptr = wide_ptr.ptr as *mut T;
            for i in 0..wide_ptr.len {
                std::ptr::drop_in_place(data_ptr.offset(i));
            }
```

        }
    }
}
```

## C++ like vtables

Most of the boilerplate is the same as with regular vtables.

```rust
unsafe impl<T: MyTrait> const CustomUnsize<dyn MyTrait> for T {
    fn unsize<const owned: bool>(ptr: *mut T) -> CppPtr {
        unsafe {
            let new = Box::new((default_vtable::<T>(), std::ptr::read(ptr)));
            std::alloc::dealloc(ptr);

            CppPtr(Box::into_ptr(new) as *mut _)
        }
    }
}

struct CppPtr(*mut (Vtable, ()));

unsafe impl CustomProjection for CppPtr {
    unsafe fn project(ptr: CppPtr, offset: usize) -> *mut u8 {
        (&raw mut (*ptr.0).1).offset(offset)
    }
}

// No CustomProjectionToUnsized as you can't have
// `struct Foo(i32, dyn MyTrait);` as that would require
// us to rewrite the vtable on unsizing. Rust puts the
// unsized field at the end, while C++ puts the in the front of
// the class.

unsafe impl CustomUnsized for dyn MyTrait {
    type WidePointer = CppPtr;
    fn size_of(ptr: CppPtr) -> usize {
        unsafe {
            (*ptr.0).0.size
        }
    }
    fn align_of(self: CppPtr) -> usize {
        unsafe {
            (*ptr.0).0.align
        }
    }
}
```

> ## Slicing into matrices via the `Index` trait

Reviewer: This is such a cool example.

we have to invoke a function, such as `array.index(s![5..8, 3..])` to obtain 
a slice that takes indices 5-7 in the first dimension and all indices after
the 3rd in the second dimension.
Instead of having our own `ArrayView` type, we can create a trait with a 
custom vtable and use the `Index` trait. We reuse the `ArrayView` type
as the type of the wide pointer.

```rust
trait Slice2<T> {}

unsafe impl<T> CustomUnsized for dyn Slice2<T> {
    type WidePointer = ndarray::ArrayView<T, Ix2>;
    fn size_of(ptr: WidePointer) -> usize {
        ptr.len() * std::mem::size_of::<T>()
    }
    fn align_of(ptr: WidePointer) -> usize {
        std::mem::align_of::<T>()
    }
}

impl<'a, T, U, V> Index<&'a SliceInfo<U, V>> for Array2<T> {
    type Output = dyn Slice2<T>;
    fn index(&self, idx: &'a SliceInfo<U, V>) -> &dyn Slice2<T> {
        unsafe {
            // This can get a better impl, but for simplicity we reuse
            // the existing function.
            transmute(Array2::index(idx))
        }
    }
}
```

## Zero sized references to MMIO

Instead of having one type per MMIO register bank, we could have one

> trait per bank and use a zero sized wide pointer format. There's no `Unsize`

Author: This is just one of the design scopes. Right now there are already setups where ppl do this, but they do it with one ZST struct per bank and methods on those. I don't want to get into the merit of this at all, even if I have strong opinions on it. This is just to show *that* you can do it

> trait per bank and use a zero sized wide pointer format. There's no `Unsize`

Reviewer: It's simply a design parameter that significantly complicates the design space, so I think the RFC would be improved if it was strongly justified.

}

unsafe trait CustomProjectionToUnsized: CustomUnsized {
    /// The offset is in bytes and must be the exact offset from the start of the unsized struct to its unsized field.
    unsafe fn project_unsized(ptr: <Self as CustomUnsized>::WidePointer, offset: usize) -> <Self as CustomUnsized>::WidePointer;
}
```

The 

# Drawbacks
[drawbacks]: #drawbacks

* This may be a serious case of overengineering. We're basically taking vtables out of the language and making dynamic dispatch on trait objects a user definable thing.
* This may slow down compilation, likely entirely preventable by keeping a special case in the compiler for regular trait objects.
* This completely locks us into never adding multiple vtable formats for a single trait. So you can't use a trait both as a C++ like vtable layout in some situations and a Rust wide pointer layout in others.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

This is a frequently requested feature and as a side effect obsoletes `extern type`s which have all kinds of problems (like the inability to invoke `size_of_val` on pointers to them).

> 

Reviewer: Consider mentioning that we could instead let the language handle just the data pointer, ridding ourselves of the project and unsize functions at the cost of not controlling the entire pointer layout.

Author: In that case we'd go with any of the other RFCs ;) Which is something I do list further down.

```rust
#[repr(C)]
struct Pointer {
    ptr: *mut u8,
    vtable: &'static VTable,
}

/// For going from a `&SomeStruct<dyn Trait>` to a field of `SomeStruct`.
unsafe impl CustomProjection for Pointer {
    unsafe fn project(ptr: Pointer, offset: usize) -> *mut u8 {
        ptr.ptr.offset(offset),
    }
}

unsafe impl CustomProjectionToUnsized for Pointer {
    /// For going from `&SomeStruct<dyn Trait>` to the unsized field
    unsafe fn project_unsized(ptr: Pointer, offset: usize) -> Pointer {

>         Pointer {

Reviewer 7: Isn't this basically just returning a copy of `ptr`? Is that what it is supposed to be?

Author: oops, thanks!
